float8_experimental/float8_tensor.py

```diff
@@ -30,27 +56,55 @@
     defaults=[False, False, False, False],
 )
 
+# The object below exists for convenience, to allow Float8Tensor to use
+# the right config based on which gemm from `y`, `dL_dX`, `dL_dW` is
+# being called.
+LinearMMConfig = namedtuple(
+    "LinearMMConfig",
+    ["y", "dL_dX", "dL_dW"],
+    defaults=[
+        ScaledMMConfig(False, True, False, False),
+        ScaledMMConfig(False, False, False, False),
+        ScaledMMConfig(False, False, False, False),
+    ],
+)
 
-def merge_mm_configs(
-    a_mm_config: ScaledMMConfig, b_mm_config: ScaledMMConfig
-) -> ScaledMMConfig:
-    """Merges two mm_configs together emulate behavior must match,
-    However we want to use_fast_accum in forward and not in backward.
-    We do this by populating the fields of the backproping grad. Same applies for fp8_output.
 
-    For both use_fast_accum and fp8_output, if either config is False, the merged config will be False.
-    """
-    assert (
-        a_mm_config.emulate == b_mm_config.emulate
-    ), "Both mm_configs must have the same emulate value, but got {} and {}".format(
-        a_mm_config.emulate, b_mm_config.emulate
-    )
-    return ScaledMMConfig(
-        emulate=a_mm_config.emulate,
-        use_fast_accum=a_mm_config.use_fast_accum and b_mm_config.use_fast_accum,
-        fp8_output=a_mm_config.fp8_output and b_mm_config.fp8_output,
-        pad_inner_dim=a_mm_config.pad_inner_dim and b_mm_config.pad_inner_dim,
-    )
+# Given a Float8Tensor, the enum below describes the expected role of this
+# tensor in the three gemms present in the fw + bw pass of a Linear layer.
+# This is used to choose the right config for a float8 gemm when the
+# gemm is performed.
+class GemmInputRole(enum.Enum):
+    X = "x"
+    W = "w"
+    DL_DY = "dL_dY"
+
+
+# choose which scaled_mm_config to use based on gemm inputs
+def choose_scaled_mm_config(
+    a_role: GemmInputRole,
+    a_linear_mm_config: LinearMMConfig,
+    b_role: GemmInputRole,
+    b_linear_mm_config: LinearMMConfig,
+):
+    if a_role is GemmInputRole.X and b_role is GemmInputRole.W:
+        assert (
+            a_linear_mm_config.y == b_linear_mm_config.y
+        ), f"linear_mm_config.y mismatch: {a_linear_mm_config.y} vs {b_linear_mm_config.y}"
+        return a_linear_mm_config.y
+    elif a_role is GemmInputRole.DL_DY and b_role is GemmInputRole.W:
+        assert (
+            a_linear_mm_config.dL_dX == b_linear_mm_config.dL_dX
+        ), f"linear_mm_config.dL_dX mismatch: {a_linear_mm_config.dL_dX} vs {b_linear_mm_config.dL_dX}"
+        return a_linear_mm_config.dL_dX
+    else:
```

Comment:
nit: I think if you just made 1 more elif branch for the "correct role" and then an else path for the error case would be a little easier to read